vimiv/imutils/metadata.py

```diff
@@ -0,0 +1,284 @@
+# vim: ft=python fileencoding=utf-8 sw=4 et sts=4
+
+# This file is part of vimiv.
+# Copyright 2017-2023 Christian Karl (karlch) <karlch at protonmail dot com>
+# License: GNU GPL v3, see the "LICENSE" and "AUTHORS" files for details.
+
+"""Utility functions and classes for metadata handling.
+
+This module provides a common interface for all metadata related functionalities.
+`MetadataHandler` is used to interact with the metadata of the current image. It
+relies on the functionality provided by optionally loaded metadata plugins. Such
+plugins implements the `MetadataPlugin` abstract class and registers that class
+using the `register` function.
+
+Module Attributes:
+    _registry: List of registered `MetadataPlugin` implementations.
+"""
+
+from abc import ABC, abstractmethod
+import contextlib
+import itertools
+from typing import Dict, Tuple, NoReturn, Sequence, Iterable, Type, List
+
+from vimiv.utils import log
+
+_logger = log.module_logger(__name__)
+
+# Type returned by `MetadataHandler.get_metadata`.
+# Key is the metadata key. Value is a tuple of descriptive name and value for that key.
+MetadataDictT = Dict[str, Tuple[str, str]]
+
+
+class MetadataPlugin(ABC):
+    """Abstract class implemented by plugins to provide metadata capabilities.
+
+    Implementations of this class are required to overwrite `__init__`, `name`,
+    `version`, `get_metadata` and `get_keys`.
+    The implementations of `copy_metadata` and `get_date_time` is optional.
+    """
+
+    @abstractmethod
+    def __init__(self, _path: str) -> None:
+        """Initialize metadata handler for a specific image.
+
+        Args:
+            _path: Path to current image.
+        """
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """Get the name of the used backend.
+
+        If no backend is used, return the name of the plugin.
+        """
+
+    @property
+    @abstractmethod
+    def version(self) -> str:
```

Comment:
Fully agree with making these abstract classemethods or staticmethods, leaving a comment so we have a place to tick it off :blush:

Reply:
Changed in e16b7632a61cbf8b4e5617038f9d76836df7fd8e